### ISSUE TITLE
"Object of type date is not JSON serializable" with `Package: true`

### ISSUE THREAD

**Issue by mkielar** [Kotaimen/awscfncli#67]
cfn-cli, version 2.1.2
Python, version 3.7.1

When I try to run:
```
cfn-cli -f ./stack.yaml -a my-cf-bucket stack sync  --confirm
```
this happens:
```
Syncing stack Default.Stack
StackName: <edited>
Profile: <edited>
Region: eu-west-2
Account: <edited>
Identity: <edited>
Using specified artifact bucket s3://<edited>
Successfully packaged artifacts and uploaded to s3://<edited>.
Traceback (most recent call last):
  File "c:\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Python37\Scripts\cfn-cli.exe\__main__.py", line 9, in <module>
  File "c:\python37\lib\site-packages\awscfncli2\__main__.py", line 31, in main
    auto_envvar_prefix='CFN'
  File "c:\python37\lib\site-packages\click\core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "c:\python37\lib\site-packages\click\core.py", line 717, in main
    rv = self.invoke(ctx)
  File "c:\python37\lib\site-packages\click\core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\python37\lib\site-packages\click\core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "c:\python37\lib\site-packages\click\core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\python37\lib\site-packages\click\core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "c:\python37\lib\site-packages\click\decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "c:\python37\lib\site-packages\awscfncli2\cli\utils\deco.py", line 16, in wrapper
    return f(ctx, *args, **kwargs)
  File "c:\python37\lib\site-packages\awscfncli2\cli\stack\sync.py", line 41, in sync
    ctx.obj.runner.run(command)
  File "c:\python37\lib\site-packages\awscfncli2\runner\runbook\base.py", line 48, in run
    command.run(context)
  File "c:\python37\lib\site-packages\awscfncli2\runner\commands\stack_sync_command.py", line 39, in run
    stack_context.run_packaging()
  File "c:\python37\lib\site-packages\awscfncli2\runner\runbook\boto3_context.py", line 105, in run_packaging
    prefix=self.parameters['StackName']
  File "c:\python37\lib\site-packages\awscfncli2\runner\runbook\package.py", line 95, in package_template
    separators=(',', ':'))
  File "c:\python37\lib\json\__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "c:\python37\lib\json\encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "c:\python37\lib\json\encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "c:\python37\lib\json\encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type date is not JSON serializable
```

The problem seems to be with this section in `awscfncli2\runner\runbook\package.py`:
https://github.com/Kotaimen/awscfncli/blob/master/awscfncli2/runner/runbook/package.py#L92-L95

If I add `default=str` to that `json.dumps` call, it continues past that point, but I'm not sure what else am I breaking with this...

**Comment by mkielar** [comment]
I've added some logging, and while I cannot show you the full contents of the `template_body` variable, the only thing that resembles a `date` object is this:
```
"AWSTemplateFormatVersion": "2010-09-09",
```
That'd suggest this will fail for all stacks with `Package: true`...

**Comment by Kotaimen** [comment]
Hi @mkielar,

Could you show us the awscli/boto3 version you are using?
You can adding -vv to show more debug logs (please remove sensitive information).

Hi @GlieseRay is it possible they this is related to stack reference?

**Comment by mkielar** [comment]
Hi,

the versions:
```
awscfncli        0.5.2
awscfncli2       2.1.2
awscli           1.16.92
boto             2.49.0
boto3            1.9.82
botocore         1.12.82
```
However, I digged a bit deeper, and think I know what caused the issue:
This was my Cloudformation template:
```
AWSTemplateFormatVersion: 2010-09-09
Description: "Some stack"

Parameters:
  ...
```
You can see, that the `AWSTemplateFormatVersion` value was not enclosed in quotes, which YAML parsers treat as Date, not String. This is somehow against the specs, as [this states explicitly that the value must me a String](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/format-version-structure.html). 

Running `cloudformation package` on such stack however worked correctly, and did not really mind it being a date. As I'm currently migrating from `awscfncli 0.5.2` (where I used `cloudformation package` followed by `cfn changeset create/execute`, and it worked with the date value as well), I was a bit surprised to see that it doesn't work with `2.x` with `cfn-cli sync ...`.

Anyway, after fixing the `AWSTemplateFormatVersion` values to be actual Strings, `2.x` is happy again, so I'm leaving this up to you if you want to fix this to behave more like pure `aws cloudformation ...` calls, or close the ticket and stick with the spec.

**Comment by mkielar** [comment]
The same issue appears when I try to create a PolicyDocument, or put the date value as date in any other field. E.g.:
```
  DedicatedSnsTopicPolicy:
    Type: 'AWS::IAM::ManagedPolicy'
    Properties:
      Description: Access for app
      Path: /
      ManagedPolicyName: 'MyPolicyName'
      PolicyDocument:
        Version: 2012-10-17
        Statement:
          Effect: Allow
          Action: sns:Publish
          Resource: !Ref SnsTopic
```

In this case the `Version` property is causing the failure, and if I wrap it with quotes, it works.

**Comment by Kotaimen** [comment]
Hi @mkielar,
If you don't wrap `Version` in string, YAML treats it as a date object and loads it back as python's `Date` object. 

If `package` is `false` `cfn-cli` is handling the template itself and just treat the whole as text.
But if `package` is `true` `cfn-cli` will call boto3's internal CloudFormation packaging to do the heavy lifting and for the API they provided returns a loaded Python object:

<img width="521" alt="image" src="https://user-images.githubusercontent.com/1639478/51538026-20239100-1e8b-11e9-896e-a2485fb7a829.png">

So that's why your template failed to deploy, please quote ambiguous values in YAML and try again.

In CloudFormation documentation AWS always quoting dates (and boolean values):
``` YAML
CreateTestDBPolicy: 
  Type: AWS::IAM::ManagedPolicy
  Properties: 
    Description: "Policy for creating a test database"
    Path: "/"
    PolicyDocument: 
      Version: "2012-10-17"
      Statement: 
        - 
          Effect: "Allow"
          Action: "rds:CreateDBInstance"
          Resource: 
```

@GlieseRay We need to check whether this causes problem in official `aws` command, if `awscli` hides this error to user by convert everything to string then we also need to do that.

**Comment by Kotaimen** [comment]
Hi, @GlieseRay:
This is what `awscli` does:

````awscli/customizations/cloudformation/yamlhelper.py```

```python
def yaml_dump(dict_to_dump):
    """
    Dumps the dictionary as a YAML document
    :param dict_to_dump:
    :return:
    """
    return yaml.dump(
        dict_to_dump,
        default_flow_style=False,
        Dumper=FlattenAliasDumper,
    )
```

So we shall replicate this method and also fix this issue at our side.

**Comment by Kotaimen** [comment]
Hi @mkielar, 
I fixed that in develop and waiting for @GlieseRay cherrypick into a new hot fix release, thank for your patience and great help!

**Comment by GlieseRay** [comment]
Hi @mkielar , It's been fixed and released in 2.1.4. Please check it.
